[WIP] CLJC port

plexus opened the pull request.

evinasgu committed:
Changing extension from *.clj to *.cljc

evinasgu committed:
Adding reader conditionals at dispatch namespace

evinasgu committed:
Configuring project.clj for cljs

evinasgu committed:
modifying exceptions and java calls. Not finished yet

evinasgu committed:
Adding cljs-time to project.clj and changes on printer.cljc

evinasgu committed:
Finishing changes in printer namespace. Approbation pending

evinasgu committed:
More changes incoming for Clojure Types

evinasgu committed:
Committing for lambdaisland remote

evinasgu committed:
Finished the changes to printer.cljc. Require command is working

plexus committed:
Remove .cljs_node_repl

plexus committed:
Add deps.edn / tests.edn setup

plexus commented on project.clj:
```diff
@@ -10,7 +10,36 @@
+   [com.andrewmcveigh/cljs-time "0.5.2"]]
+
+  :profiles {:dev {:plugins [[lein-difftest "2.0.0"]
+                             [com.cemerick/clojurescript.test "0.3.3"]
```
Please use Kaocha for testing, we should eat our own dogfood :) I added a deps.edn / tests.edn / bin/kaocha, so you can run

```
bin/kaocha clj
bin/kaocha cljc
```

evinasgu replied:
Ok! I will start here today :) I have some tests pending so it's a good opportunity to take this as well

plexus commented on src/puget/dispatch.cljc:
```diff
@@ -18,10 +18,11 @@
-       (throw (IllegalArgumentException.
-                "chained-lookup must be provided at least one dispatch function to try.")))
+       #?(:clj (throw (IllegalArgumentException. no-chain-lookup-provided-message)) ; I don't like this. Candidate for improvement
+          :cljs (throw no-chain-lookup-provided-message)))
```
How about we turn this into an `(ex-info ...)`? this way you get an exception that is cross-platform.

evinasgu replied:
Got it

plexus commented on src/puget/dispatch.cljc:
```diff
@@ -93,19 +93,23 @@
   (fn lookup
```
We're going to have to rethink how some of this works. The strategy in Clojure for inheritance-lookup seems to be: build up a list of superclasses, protocols, and interfaces, and then look for them one by one in the given dispatch map.

In ClojureScript we can't do that. There's some fuzzy concept that is similar to a superclass (through prototype.contructor, or if it's using goog.inherits then there's superClass_.constructor), but there is no way to get a list of all protocols that a given type implements. We can only do it the other way around. Given a certain prototype, does `(satisfies? proto x)`. But that requires a concrete value, not a class/type/constructor...

plexus commented on src/puget/printer.cljc:
```diff
@@ -218,7 +261,7 @@
   [printer value]
   (let [lookup (:print-handlers printer)
-        handler (and lookup (lookup (class value)))]
+        handler (and lookup (lookup (get-type value)))]
```
This is the point where the Clojure and ClojureScript story are going to start diverging. In Clojure `lookup` is a function of type `java.lang.Class -> clojure.lang.IFn`, i.e. it turns a type/class into a handler function.

This lookup function is built up using the helpers in `dispatch`, especially `inheritance-lookup`, this last one is going to have to work differently in ClojureScript.

- Take a value (not a type)
- look for an entry in the dispatch map where `(= t (type value))`, i.e. a literal match of the concrete type
- ~~if that doesn't work, look for an entry where `(satisfies? t value)`~~ => but this won't work either, because satisfies? is a macro... I need to do some research here, see what out options are.
- Finally if we still haven't found something then look for an Object entry.

evinasgu replied:
After some intents and "debugging" a while I think I got this point. Adding more to this context, accessing parent structure in JS looks like an intricate task, especially from ClojureSript. This will impact on inheritance-lookup as well. I think it is possible but not natural. 
Maybe we can start mapping keys associated with keys and trigger some specific behavior from the dispatcher side. This is my ClojureScript beginner perspective though...

evinasgu committed:
handling .cljs_node_repl folder

evinasgu committed:
pulling changes

plexus committed:
ClojureScript: Use predicates instead of type dispatch

JavaScript does not have a uniform way of traversing a type hierarchy, or for
dynamically checking if a value satisfies a given type. (instance? and friends
are macros that only work with fixed values.)

Instead switch to an approach based on predicates.

evinasgu committed:
Added files to start working with tests

evinasgu committed:
Added files to start working with tests

evinasgu committed:
Merge branch 'handlers-work' of github.com:lambdaisland/puget into handlers-work

evinasgu committed:
Adding needed exclusions

evinasgu committed:
Added tests for dispatch namespace - ClojureScript support

evinasgu committed:
Adding tests for printer namespace - ClojureScript support

evinasgu committed:
Finishing tests - Clojure Support

plexus commented on src/puget/macros.clj:
```diff
@@ -0,0 +1,4 @@
+(ns puget.macros)
+
+(defmacro current-ns []
+  (str *ns*))
```
This seems like it isn't used?

evinasgu replied:
Removed

plexus commented on src/puget/dispatch.cljc:
```diff
@@ -0,0 +1,121 @@
+  as a lookup function."
+  (:require
+   [clojure.string :as str]
+   #?(:cljs [goog.string :as gstring])))
```
Unused require.

plexus commented on src/puget/printer.cljc:
```diff
@@ -84,15 +84,55 @@
+  "Verify if a promise is resolved"
+  [promise]
+  #?(:clj (future-done? promise)
+     :cljs (js/Promise.resolved? promise)))
```
This can't be right, a native JavaScript function can't have a `?` in the name.

evinasgu replied:
I will add a test for this method as well

plexus commented on src/puget/printer.cljc:
```diff
@@ -112,7 +152,7 @@
     :tag       [:red]
 
-    ; primitive values
+                                        ; primitive values
```
Here the original code uses a single `;` for comments, which in LISP convention is only used for end of line comments, which is why your (or mine) editor moved them here. Best to change these to `;;` and move them to their original column.

plexus commented on src/puget/printer.cljc:
```diff
@@ -84,15 +84,55 @@
+  #?(:clj (System/identityHashCode o)
+     :cljs (hash o)))
+
+
```
Pay attention to little cleanliness things like this. A single newline, not two.

plexus commented on src/puget/printer.cljc:
```diff
@@ -264,70 +305,95 @@
+                      date-formatter (cljs-time.format/formatter "yyyy-MM-dd'T'HH:mm:ss.SSS-00:00")]
+                  (cljs-time.format/unparse date-formatter dt)))))
+
+
```
Unnecessary extra newline.

plexus commented on src/puget/printer.cljc:
```diff
@@ -264,70 +305,95 @@
+         (format-unknown printer value "Class" (get-type-name value)))
+
+
+
```
newlines.

plexus commented on src/puget/printer.cljc:
```diff
@@ -264,70 +305,95 @@
+       (tagged-handler
+        'object
+        (fn [x]
+          (js->clj x)))})))
```
Use `'js` for the tag here, that way the output will look like `#js {"foo" "bar"}`, which is compatible with the ClojureScript reader.

evinasgu replied:
Perfect I will replace 'object by 'js tag

plexus commented on src/puget/printer.cljc:
```diff
@@ -339,26 +405,28 @@
+     :cljs (dispatch/chained-lookup
+            (dispatch/predicate-lookup platform-handlers)
+            (dispatch/predicate-lookup clojure-handlers)
+            (dispatch/predicate-lookup clojure-interface-handlers))))
```
:+1: looks good

plexus commented on src/puget/printer.cljc:
```diff
@@ -339,26 +405,28 @@
 
 (defrecord CanonicalPrinter
-  [print-handlers]
+    [print-handlers]
```
Seems this is one of those things where editors disagree on how to indent this. Best to move the `[print-handlers]` to the end of line 424.

plexus commented on src/puget/printer.cljc:
```diff
@@ -339,26 +405,28 @@
   fv/IVisitor
 
-  ; Primitive Types
+                                        ; Primitive Types
```
Same comment as above, change to `;;` and re-indent.

plexus commented on src/puget/printer.cljc:
```diff
@@ -468,30 +540,31 @@
   fv/IVisitor
 
-  ; Primitive Types
+                                        ; Primitive Types
```
change to `;;` and re-indent

plexus commented on src/puget/printer.cljc:
```diff
@@ -573,22 +646,22 @@
                            [:span
                             (format-doc this k)
-                            (if (coll? v)
+                            (if (coll? v)                              
```
trailing whitespace? configure your editor to remove trailing whitespace if possible. In emacs

```
(add-hook 'before-save-hook #'delete-trailing-whitespace)
```

evinasgu replied:
Great! I will add the hook to my configuration

plexus commented on src/puget/printer.cljc:
```diff
@@ -573,22 +646,22 @@
+                            (if (coll? v)                              
                               map-coll-separator
-                              " ")
+                              " ") ;; Original value " "
```
what does this comment mean?

plexus commented on src/puget/printer.cljc:
```diff
@@ -573,22 +646,22 @@
             map-doc [:group
                      (color/document this :delimiter "{")
-                     [:align (interpose [:span map-delimiter :line] entries)]
+                     [:align (interpose [:span map-delimiter :line] entries)]     
```
trailing whitespace

plexus commented on src/puget/printer.cljc:
```diff
@@ -573,22 +646,22 @@
 
 
-  ; Clojure Types
+                                        ; Clojure Types
```
change to `;;` and re-indent.

plexus commented on src/puget/printer.cljc:
```diff
@@ -613,12 +686,12 @@
 
 
-  ; Special Types
+                                        ; Special Types
```
FIXME

plexus commented on test/puget/cljs/ansi_test.cljs:
```diff
@@ -0,0 +1,134 @@
+(def c-class-delimiter [:blue])
+(def c-class-name      [:white])
+
+
```
Stick to single newlines.

plexus commented on test/puget/cljs/ansi_test.cljs:
```diff
@@ -0,0 +1,134 @@
+(ns puget.cljs.ansi-test
+  (:require
+    [clojure.string :as str]
```
Put the first namespace on the same line as the require.

plexus commented on test/puget/cljs/ansi_test.cljs:
```diff
@@ -0,0 +1,134 @@
+(ns puget.cljs.ansi-test
+  (:require
+    [clojure.string :as str]
+    [clojure.test :as t]
```
See the styleguide about this, basically either use `[clojure.test :refer :all]` (if it's clj only), or `[clojure.test :refer [...]]` if it's cljs/cljc.

plexus commented on test/puget/cljs/ansi_test.cljs:
```diff
@@ -0,0 +1,134 @@
+  (-> s
+      (str/replace #"\u001b\[([0-9;]*)[mK]" "፨⦕$1⦖፨")
+      (str/split #"፨")
+      (->> (filter seq))))
```
note that we actually have a library named [ansi](https://github.com/lambdaisland/ansi) for parsing these things into clojure data structures. Could be useful here for testing.

plexus commented on test/puget/cljs/ansi_test.cljs:
```diff
@@ -0,0 +1,134 @@
+                               ["item14"
+                                ["item15"]]]]]]]]]]]]]]]]
+    (t/is (= (escape-ansi
+             (str (ansi "[" c-delimiter)
```
indentation is off here.

evinasgu committed:
Adding style related modifications

evinasgu committed:
Fixing style errors, Adding printer tests and modified README.md

evinasgu committed:
Adding ex-info for exceptions and fixed warning for clojure.walk/postwalk unused Var

evinasgu commented:
I think this is ready for merge :)

plexus committed:
Avoid js->clj as we do not want the deep recursive conversion
